actors/builtin/market/market_state.go

```diff
@@ -205,200 +205,204 @@ func (st *State) generateStorageDealID() abi.DealID {
-		st.TotalProviderLockedCollateral = big.Sub(st.TotalProviderLockedCollateral, amount)
-	}
+	usePendingDeals bool
+	pendingDeals    *adt.Map
 
-	return nil
+	useDealsByEpoch bool
+	dealsByEpoch    *SetMultimap
 }
 
-// move funds from locked in client to available in provider
-func (st *State) transferBalance(rt Runtime, fromAddr addr.Address, toAddr addr.Address, amount abi.TokenAmount, et, lt *adt.BalanceTable) {
-	Assert(amount.GreaterThanEqual(big.Zero()))
+func newMarketStateMutation(st *State, store adt.Store) *marketStateMutation {
+	return &marketStateMutation{st: st, store: store}
+}
 
-	if err := et.MustSubtract(fromAddr, amount); err != nil {
-		rt.Abortf(exitcode.ErrIllegalState, "subtract from escrow: %v", err)
+func (m *marketStateMutation) build() (*marketStateMutation, error) {
+	if m.useDealProposals {
+		proposals, err := AsDealProposalArray(m.store, m.st.Proposals)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load deal proposals: %w", err)
+		}
+		m.dealProposals = proposals
 	}
 
-	if err := st.unlockBalance(lt, fromAddr, amount, ClientStorageFee); err != nil {
-		rt.Abortf(exitcode.ErrIllegalState, "subtract from locked: %v", err)
+	if m.useDealStates {
+		states, err := AsDealStateArray(m.store, m.st.States)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load deal state: %w", err)
+		}
+		m.dealStates = states
 	}
 
-	if err := et.Add(toAddr, amount); err != nil {
-		rt.Abortf(exitcode.ErrIllegalState, "add to escrow: %v", err)
+	if m.useLockedTable {
+		lt, err := adt.AsBalanceTable(m.store, m.st.LockedTable)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load locked table: %w", err)
+		}
+		m.lockedTable = lt
 	}
-}
 
-func (st *State) slashBalance(et, lt *adt.BalanceTable, addr addr.Address, amount abi.TokenAmount, reason BalanceLockingReason) error {
-	Assert(amount.GreaterThanEqual(big.Zero()))
+	if m.useEscrowTable {
+		et, err := adt.AsBalanceTable(m.store, m.st.EscrowTable)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load escrow table: %w", err)
+		}
+		m.escrowTable = et
+	}
 
-	if err := et.MustSubtract(addr, amount); err != nil {
-		return xerrors.Errorf("subtract from escrow: %v", err)
+	if m.usePendingDeals {
+		pending, err := adt.AsMap(m.store, m.st.PendingProposals)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load pending proposals: %w", err)
+		}
+		m.pendingDeals = pending
+	}
+
+	if m.useDealsByEpoch {
+		dbe, err := AsSetMultimap(m.store, m.st.DealOpsByEpoch)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load deals by epoch: %w", err)
+		}
+		m.dealsByEpoch = dbe
 	}
 
-	return st.unlockBalance(lt, addr, amount, reason)
+	return m, nil
 }
 
-////////////////////////////////////////////////////////////////////////////////
-// Method utility functions
-////////////////////////////////////////////////////////////////////////////////
+func (m *marketStateMutation) withDealProposals() *marketStateMutation {
+	m.useDealProposals = true
```

Comment:
If you run with the permisssions type idea you could include `Invalid` as a default zero valued type serving the same purpose as boolean false zero value in these options and replace the boolean fields with permission type fields.

Reply:
Making this change.